### ISSUE TITLE
Support gradual re-import of database

### ISSUE THREAD

**Issue by eddieantonio** [UAlbertaALTLab/recording-validation-interface#101]
Sometimes the metadata changes, or new recordings are added, or annotations are modified.

Currently, in order to re-import the recordings, we have to:

 - delete the database
 - rerun the importer from scratch

Here me out though: what if we told the importer to only re-import what has _changed_?

**Comment by nienna73** [comment]
Is this something that Django supports and we're missing a command in that process?
Or do we have to modify our own `importrecordings` command?

**Comment by eddieantonio** [comment]
I don't believe there is Django magic for this; we need to manually scan the recordings + metadata to see if they're different than what's in the database, then update _just_ the data that has changed.

I would suggest refactoring the import process into two stages: one stage scans the metadata and filesystem for potentially data to import; then the second stage actually opens the files, parses the data, extracts and transcodes the audio, and stores it in the database.

After that refactoring is done, you could insert a third stage that prunes the results from stage one ("scanning") to only apply the last stage to data that has _changed_. Does this make sense?

**Comment by nienna73** [comment]
If I understand correctly, this task includes a refactor of the current process to first find the data to import and second, import the data. Then the gradual import comes in afterwards. Is this correct?

What constitutes "different" when we're looking for new or updated data to import?

**Comment by eddieantonio** [comment]
> If I understand correctly, this task includes a refactor of the current process to first find the data to import and second, import the data. Then the gradual import comes in afterwards. Is this correct?

Correct!

> What constitutes "different" when we're looking for new or updated data to import?

I'm not really sure! I think if a SINGLE BIT changes in any of the audio or annotation files, then that session should be marked as changed. Otherwise, consider the session as "imported". Does this make sense?

**Comment by nienna73** [comment]
Yes, I think that makes sense. These are the things that I think can change from one session to another:

- Metadata sheet
- .eaf file
- .wav file

Is there anything else?

**Comment by eddieantonio** [comment]
I can't think of anything else 😅

**Comment by nienna73** [comment]
So far, I'm checking if the modify date on the .eaf file is more recent than the modify date on the DB, and doing the same with the .wav files. This should cover most cases, but feels too simple to work for all cases.

Checking if the metadata has changed is a little more complex and I can't quite wrap my head around how to check this info yet. I'll keep thinking on it.

I also import everything if the DB was created on the same day the extraction process is running.

**Comment by andrewdotn** [comment]
A classic way to handle the metadata update: hash the metadata, and save the hash in the database. Then you have a quick way to check if the metadata has changed.

    def metadata_hash(metadata):
        return hashlib.sha256(json.dumps(metadata).encode('UTF-8')).hexdigest()

**Comment by nienna73** [comment]
I found a scenario where using the date for the .eaf and .wav files won't be reliable:

1. Create DB
2. Add a new .eaf file
3. Someone validates an entry, therefore changing the DB modify time
4. Do another import, and the new .eaf file is missed

Can this be fixed by using the create time for the DB, or do I need to hash all the thing?

**Comment by nienna73** [comment]
Hmm okay, also: do I need one hash for all the metadata entries? or a hash per entry? (English is hard)
I was hoping I could pinpoint which session/s had new metadata and only import that/those one/s, thus associating a metadata hash with each session in the DB.

**Comment by nienna73** [comment]
I got excited/distracted and forgot to do the first step: refactor.
Let me refocus and do that, then we can resume this discussion when I'm less confused about where we should be checking if the data are different.

**Comment by andrewdotn** [comment]
Ah, good question. For each file, you’d store the hash of the corresponding row from the metadata spreadsheet.

**Comment by eddieantonio** [comment]
> Can this be fixed by using the create time for the DB, or do I need to hash all the thing?

![HASH ALL THE THINGS](https://user-images.githubusercontent.com/2294397/108269625-7ecf9d80-712b-11eb-9993-293141c2b5f0.jpg)

**Comment by nienna73** [comment]
@eddieantonio I think we need to pick this issue up again.

Once we've investigated all the cases we can, we'll need to do another DB import. To avoid the problem we had last time, when we had to remove a bunch of duplicates from the DB, I think we'll want a way to update (i.e. delete then re-import) existing entries.

**Comment by nienna73** [comment]
Thinking back on this, I think we were vastly over-thinking the problem. Sessions that need to be re-imported can be symlinked to a new directory and we can point the recording extraction process to that directory. Entries that already exist and haven't been updated by a user (i.e. changed the transcription, translation, and/or speaker code), can simply be replaced in the import process. This will still require some manual work, but I honestly think it's less hassle than trying to figure this out computationally.
